expression/collation.go

```diff
@@ -189,12 +168,102 @@ func deriveCoercibilityForColumn(c *Column) Coercibility {
+	case ast.Like:
+		dstCharset, dstCollation, _, err = CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args[0], args[1])
+		if err != nil {
+			return
+		}
+		return dstCharset, dstCollation, CoercibilityCoercible, err
+	case ast.In:
+		if args[0].GetType().EvalType() == types.ETString {
+			return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, types.ETInt, args...)
+		}
+	case ast.DateFormat, ast.TimeFormat:
+		charsetInfo, collation := ctx.GetSessionVars().GetCharsetInfo()
+		return charsetInfo, collation, args[1].Coercibility(), nil
+	case ast.Cast:
+		// we assume all the cast are implicit.
+		if retType == types.ETString {
+			charsetInfo, collation := ctx.GetSessionVars().GetCharsetInfo()
+			return charsetInfo, collation, args[0].Coercibility(), nil
+		}
+		return charset.CharsetBin, charset.CollationBin, args[0].Coercibility(), nil
+	case ast.Case:
+		// FIXME: case function aggregate collation is not correct.
+		return CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, retType, args...)
+	case ast.Database, ast.User, ast.CurrentUser, ast.Version, ast.CurrentRole, ast.TiDBVersion:
+		chs, coll := charset.GetDefaultCharsetAndCollate()
+		return chs, coll, CoercibilitySysconst, nil
+	}
+
+	if retType == types.ETString {
+		charsetInfo, collation := ctx.GetSessionVars().GetCharsetInfo()
+		return charsetInfo, collation, CoercibilityCoercible, nil
+	}
+	return charset.CharsetBin, charset.CollationBin, CoercibilityNumeric, nil
+}
+
 // DeriveCollationFromExprs derives collation information from these expressions.
+// Deprecated, use CheckAndDeriveCollationFromExprs instead.
+// TODO: remove this function after the all usage is replaced by CheckAndDeriveCollationFromExprs
 func DeriveCollationFromExprs(ctx sessionctx.Context, exprs ...Expression) (dstCharset, dstCollation string) {
 	dstCollation, dstCharset, _, _ = inferCollation(exprs...)
 	return
 }
 
+// CheckAndDeriveCollationFromExprs derives collation information from these expressions, return error if derives collation error.
+func CheckAndDeriveCollationFromExprs(ctx sessionctx.Context, funcName string, evalType types.EvalType, args ...Expression) (dstCharset, dstCollation string, err error) {
+	dstCharset, dstCollation, _, err = CheckAndDeriveCollationFromExprsWithCoer(ctx, funcName, evalType, args...)
+	return
+}
+
+// CheckAndDeriveCollationFromExprsWithCoer is the same with CheckAndDeriveCollationFromExprs, but also return Coercibility.
+func CheckAndDeriveCollationFromExprsWithCoer(ctx sessionctx.Context, funcName string, evalType types.EvalType, args ...Expression) (dstCharset, dstCollation string, coercibility Coercibility, err error) {
+	coll, chs, coercibility, legal := inferCollation(args...)
+	if !legal {
+		return "", "", CoercibilityIgnorable, illegalMixCollationErr(funcName, args)
+	}
+
+	if evalType != types.ETString && coercibility == CoercibilityNone {
```

Comment:
How do we know this is illegal?

Reply:
it is from the MySQL source code. see the function `agg_item_collations` in `sql/item.cc`